### PR TITLE
Add Postgresql CI support back

### LINKED ISSUE #35: Add PostgreSQL support back
Seems it was disabled because of #20 in https://github.com/ChristophWurst/twofactor_admin/commit/78e8250be876966e09656d56bd90803e858d2a07 but it should be fine now.

### PR DESCRIPTION
**Pull request by henry-nicolas**
Closes #35
Add Travis CI tests for pgsql

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -10,8 +10,12 @@ matrix:
     env: "DB=sqlite SERVER_BRANCH=stable17"
   - php: 7.3
     env: "DB=sqlite SERVER_BRANCH=v17.0.0beta4"
+  - php: 7.3
+    env: "DB=pgsql"
   - php: 7.2
     env: "DB=mysql"
+  - php: 7.2
```
Comment (ChristophWurst) [inline .travis.yml]: IMO one postgres job is sufficient. It's highly unlikely that php7.2 works with postgres if 7.3 doesn't or the other way around

Comment (henry-nicolas) [reply .travis.yml]: Sure I will change that, also I see there is an issue with the command postgres --version in the last travis CI run that needs to be fixed.

**Comment by henry-nicolas** [comment]
Travis found one issue when enabling the module, see below:

SQLSTATE[42704]: Undefined object: 7 ERROR:  index "primary" does not exist

The command "php -f server/occ app:enable twofactor_admin" failed and exited with 1 during .

**Comment by ChristophWurst** [comment]
The problem is dropping the primary key. This doesn't work apparently. I've recently run into the very same problem with https://github.com/nextcloud/mail/. There it was luckily the cache table, so I could safely drop everything and start from scratch. But with this app we have to find a way to keep the data.

**Comment by henry-nicolas** [comment]
Christpph,

I probably do not know NextCloud internals well enough but I see that the PK is dropped in a migration class. I expect Nexcloud is running this class on install / upgrades to update the schema this app uses alongside the application.

If that is the case, wouldn't there be an option to change the logic in that file based on the DB used ? Since this apps has no postgres user (yet), there's no migration to perform.

Or did I miss something ?

**Comment by henry-nicolas** [comment]
After some further investigation, I think I would need you to shed some light on this Christoph :)

From what I can read in Travis's log file, testValidateCorrectCode() and testValidateCorrectCodeTwice() both fails on their first call to validateCode.
The SQL statement which breaks the transactions is coming from the StateChangeRegistryUpdater where there is a disableProvider function call (line 51).

I assume that validating the code once should disable the provider for the user in the test scenario so the logic make sense (you can use the temporary code only once).
But I do not understand why this UPDATE statement failed (pasted below).

Any idea what could be wrong there? Seems that we are getting close.

```
Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'UPDATE "oc_twofactor_providers" SET "enabled" = ? WHERE ("provider_id" = ?) AND ("uid" = ?)' with params [0, "admin", "testuser2388881307009327033"]:

SQLSTATE[25P02]: In failed sql transaction: 7 ERROR:  current transaction is aborted, commands ignored until end of transaction block
```

**Comment by henry-nicolas** [comment]
Suggestion: we might need to add commit calls where IDBConnection objects are used.
That is in CodeMapper and Version2Date20180926053333

It does not matter that much with SQLIte or MySQL but I believe we need to commit at the end of the transaction in pgsql.
I assume that if we make changes on a row which was not previously commited, it might trigger this issue.

If that suggestion make sense, then I will add another commit to perform some IDBConnection::commit calls were queryBuilder is used.

Let me know :)

**Comment by henry-nicolas** [comment]
I just tested the addition of commit() calls on the IDBconnection db objects without success, it triggers more errors like "no active transaction".

It seems that the pgsql error 25P02 reported above is caused by something else, just not sure what exactly.

**Comment by ChristophWurst** [comment]
Sorry for my slow responses.

We should try to find a solution without a db type switch in the migration file. Instead we can

* Delete all existing migration files
* Create a new migration that creates a new (temporary) table
* In the *after* step we try to read from the old table, but do all kinds of error handling so this is an optional check and if anything does :boom: we carry on without the data
* In another migration we drop the old table if it exists
* In a third migration we create the actual table the way it should be and copy back the data from the temporary one
* In a fourth and final migration we delete the temporary table

Would that make sense?

**Comment by henry-nicolas** [comment]
Sure that would be cleaner that the dbtype switch solution but note that the type switch is already part of this PR and Travis tests are failling later in the code on another error.

See above DBAL Driver Exception for the details. Rewriting the Migration Scripts might provide cleaner code but I am not sure it would address this error (was kind of stuck there). 

Any idea how to fix that one? I'm not that interested in rewriting the migration scripts if later on we are still stuck with the very same DBAL Exception which I couldn't solve.

**Comment by henry-nicolas** [comment]
@ChristophWurst  Do you believe that re-writing the Migration Scripts would help solving this DBAL DriverException? 

I am concerned that this error would remain and I would not know how to fix it.

**Comment by ChristophWurst** [comment]
Yes, it would as the problematic migration code would be gone.

I have to find some time to set up a pgsql database locally, write the new migrations and test them from each of the previous schemas so we know for sure this fix works. It's low priority, though, so I'm afraid there won't be much progress soon as this is not an official Nextcloud app.

**Comment by henry-nicolas** [comment]
Chris,

The problematic code among the migration script had been already delt with earlier on in this PR with some condition statement, switching depending on db type.

The current error is raised later on in the phpunit test cases.
If you are confident that replacing the migrations scripts would adress this dbal exception, even though the exception is not raised while running the migration scripts, then yes, I would think it is worth rewriting entirely those scripts.

Please let me know, I want to make sure we are aligned before spending more time on these migration scripts.

Regards,

**Comment by SuperSandro2000** [comment]
Postgres support is essential to me also I want to be able to reset users 2fa if they stand in my office and can't login anymore. Will this be fixed?

**Comment by henry-nicolas** [comment]
@SuperSandro2000 understood and same here. Note however that the initial intent of this PR was to add back pgsql CI checks with Travis and not to fix all the potential issues coming across (I had hoped initially that the code would be compatible).

I already tried a couple of fixes on the migration scripts which looked fine but then the issues arised somewhere else.

Eventually, this is all done on my free time and it became too demanding for my to carry over this to the end.

**Comment by henry-nicolas** [comment]
All, note that I did not drop this actually, my last message was probably confusing in that regard. 

With this PR I barely had first added the CI part for PgSQL again which hepled understand that more work was needed. 
In the end, this PR started to be polluted with various commits and experiment so it is not as clean as I wish it to be.

Also, the approach I took in the earlier commits was to update the existing Migration scripts to support PostgreSql and that didn't work. After that I was kind of stucl trying to understand Nextcloud's internal as you can see here above.

Now, I am working on another PR, following @ChristophWurst  advise to completely rewrite the migration scripts with the suggested steps.
If/when I get to a positive outcome, I will be submitting another PR but with working/tested code that time :)

**Comment by henry-nicolas** [comment]
@ChristophWurst  - Eventually I completed the full rewrite for Migration Script but still the unit tests are failing at a later stage so the problem appear to be originating from somewhere else.

Details here: https://travis-ci.com/henry-nicolas/twofactor_admin/jobs/269764257
Code is here (will make another PR once it passes the CI tests): https://github.com/henry-nicolas/twofactor_admin/tree/dev

Would really help if you could have a look at the stack trace in the failled CI run and give some advise.

**Comment by ChristophWurst** [comment]
Hey :v: 
Thanks a lot for your patience. We've been very busy with the 18 release, thus I couldn't find time to look into this.

> current transaction is aborted, commands ignored until end of transaction block

IIRC on postgres you can't continue with the current transaction after an error. This is only fine in mysql and sqlite.

**Comment by henry-nicolas** [comment]
> Hey 
> Thanks a lot for your patience. We've been very busy with the 18 release, thus I couldn't find time to look into this.

Sure no problem, I understand
> 
> > current transaction is aborted, commands ignored until end of transaction block
> 
> IIRC on postgres you can't continue with the current transaction after an error. This is only fine in mysql and sqlite.

Yes, indeed, question is why there is an error in the first place?
The error occurs when the StateChangeRegistryUpdater disables the provider.

This is where I could use some advise, if you/anyone could have a look at the travis error (see below for your convenience) I believe that we could then be good with Postgresql support (the rest of the tests are successfull).

The error (full details on travis-ci):
```
1) OCA\TwoFactorAdmin\Test\Integration\CodeStorageTest::testValidateCorrectCode
Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'UPDATE "oc_twofactor_providers" SET "enabled" = ? WHERE ("provider_id" = ?) AND ("uid" = ?)' with params [0, "admin", "testuser1025769309576470205"]:

SQLSTATE[25P02]: In failed sql transaction: 7 ERROR:  current transaction is aborted, commands ignored until end of transaction block

/home/travis/build/henry-nicolas/server/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractPostgreSQLDriver.php:79
/home/travis/build/henry-nicolas/server/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:169
/home/travis/build/henry-nicolas/server/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:145
/home/travis/build/henry-nicolas/server/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:1063
/home/travis/build/henry-nicolas/server/lib/private/DB/Connection.php:220
/home/travis/build/henry-nicolas/server/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php:203
/home/travis/build/henry-nicolas/server/lib/private/DB/QueryBuilder/QueryBuilder.php:215
/home/travis/build/henry-nicolas/server/lib/private/Authentication/TwoFactorAuth/Db/ProviderUserAssignmentDao.php:91
/home/travis/build/henry-nicolas/server/lib/private/Authentication/TwoFactorAuth/Registry.php:63
/home/travis/build/henry-nicolas/server/apps/twofactor_admin/lib/Listener/StateChangeRegistryUpdater.php:51
/home/travis/build/henry-nicolas/server/lib/private/EventDispatcher/ServiceEventListener.php:76
/home/travis/build/henry-nicolas/server/3rdparty/symfony/event-dispatcher/EventDispatcher.php:247
/home/travis/build/henry-nicolas/server/3rdparty/symfony/event-dispatcher/EventDispatcher.php:73
/home/travis/build/henry-nicolas/server/lib/private/EventDispatcher/EventDispatcher.php:80
/home/travis/build/henry-nicolas/server/apps/twofactor_admin/lib/Service/CodeStorage.php:97
/home/travis/build/henry-nicolas/server/apps/twofactor_admin/tests/Integration/Service/CodeStorageTest.php:69


```

**Comment by ChristophWurst** [comment]
Let's see if https://github.com/ChristophWurst/twofactor_admin/pull/98 can help to debug this. Please update your branch once that PR got merged. Latest master is also compatible with 19, so the test are actually executed :)

**Comment by henry-nicolas** [comment]
So after the update on my branch, I re-ran the test and the tail command did not display anything, so the log is actually empty.

I will setup a test environment on my side in order to troubleshoot further what is happening here (hope to get more details than on travis ci)

**Comment by ChristophWurst** [comment]
My bad. I think we should run
```
php occ config:system:set --type int --value 0 loglevel
php occ config:system:set --type bool --value true debug
```

After the server installation. This lowers the log level to write everything and also enables debug mode. Hope this helps.

**Comment by henry-nicolas** [comment]
Hey @ChristophWurst  I noticed the following while doing some test deployments:

The below works:
`UPDATE "oc_twofactor_providers" SET "enabled" = 0 WHERE ("provider_id" = 'admin') AND ("uid" = 'test');`

The statment generated by DBAL/NextCloud looks like this and does not work:
`UPDATE "oc_twofactor_providers" SET "enabled" = 0 WHERE ("provider_id" = "admin") AND ("uid" = "test");
`

The exact message in CI is the following:
`Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'UPDATE "oc_twofactor_providers" SET "enabled" = ? WHERE ("provider_id" = ?) AND ("uid" = ?)' with params [0, "admin", "testuser1025769309576470205"]:
`

Apparently the double quotes are a source of issue, not sure how the DBAL query is translated when it hits postresql and it seems a bit odd as this is nextcloud code so used by other apps as well...

**Comment by henry-nicolas** [comment]
@ChristophWurst the tests I made with my local setup together with the debugging options you provided show the below error which seems relevant in nextcloud.log

According to the error message, it could be that the StateChangeRegistryUpdater does not work as expected because of some deprecated object type. Note that the DBAL Exception stack trace from Travis-ci is also passing through this code so it could indeed be the root cause.

Additionnaly, when performing the local testing, I do see the updates getting into the twofactor_admin_codes while the twofactor_providers table is not updated to set enabled to 0 or 1 (when a code is generated or when the code is used). So that is an extra pointer toward the StateChangeRegistryUpdater code.

The error
```
{"reqId":"nffXjcaB2yt1KhQXjnG8","level":1,"time":"2020-01-29T12:25:18+00:00","remoteAddr":"192.168.11.73","user":"test","app":"no app in context","method":"POST","url":"/login/challenge/admin","message":"Deprecated event type for OCP\\Authentication\\TwoFactorAuth\\IProvider::success: Symfony\\Component\\EventDispatcher\\GenericEvent","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15 Epiphany/605.1.15","version":"18.0.0.10"}
{"reqId":"nffXjcaB2yt1KhQXjnG8","level":0,"time":"2020-01-29T12:25:18+00:00","remoteAddr":"192.168.11.73","user":"test","app":"PHP","method":"POST","url":"/login/challenge/admin","message":"Calling the \"Symfony\\Component\\EventDispatcher\\EventDispatcherInterface::dispatch()\" method with the event name as the first argument is deprecated since Symfony 4.3, pass it as the second argument and provide the event object as the first argument instead. at /var/www/html/3rdparty/symfony/event-dispatcher/EventDispatcher.php#58","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15 Epiphany/605.1.15","version":"18.0.0.10"}

```

**Comment by henry-nicolas** [comment]
Little update here: it seems that the INSERT into the provider table works but the UPDATE does not wirth postgresql. 

UPDATE would be expected to :
- set enabled to 1 when a code is generated for a user which already used twofactor_admin (so not an INSERT)
- set enabled to 0 after code validation is succesfull
None of these 2 occurs but the INSERT works when generating a code for the first time and the user was not listed (no tuples for that user_id) in the twofactor_provider table.

I am now having a look at the ProviderUserAssignmentDao persist function which seem to handle the INSERT/UPDATE code for the TwoFactor Registry.

**Comment by ChristophWurst** [comment]
> The statment generated by DBAL/NextCloud looks like this and does not work:

What version of postgres do you use right now?

> It shows some deprecation with the objects being used in StateChangeRegistryUpdater.

You can safely ignore this warning. The events will still work.

**Comment by henry-nicolas** [comment]
On my personnal test setup, it is 11.6
On Travis, it is 9.5 (mostly because it is the version shipped with the Ubuntu LTS used on in the CI env).

In both case, the symptoms are the same: the twofactor_providers table is not updated (while insert statement are working).

**Comment by ChristophWurst** [comment]
Yes, I was just wondering if we're maybe using a version of dbal that is incompatible with your postgres. That could have explained the issue with the quotes.

**Comment by henry-nicolas** [comment]
So ... still got to test it with my on-prem NC sandbox but on Travis I managed to get all tests to pass by updating the ProviderUserAssignmentDao as you can see here:
https://github.com/henry-nicolas/server/blob/master/lib/private/Authentication/TwoFactorAuth/Db/ProviderUserAssignmentDao.php
Travis tests: https://travis-ci.com/henry-nicolas/twofactor_admin/jobs/281658782

The persist function is no longer doing a try / catch with an insert first (I suspect the insert could be what was making the whole transaction fail with postgres). Instead if uses a newly created function to check if the provider has been used for that user before (if a tuple exists or not), then it does an update or an insert based on the result.

I will run some more tests in the coming days to get this tested with my on-prem NC sandbox as well and still some finetuning that I need to do on the Migration Scripts I believe.

Meanwhile, @ChristophWurst , can you comment on the changes I made on the ProviderUserAssignmentDao.php ?

**Comment by ChristophWurst** [comment]
The problem with SELECT and then UPDATE or INSERT is prone to race conditions.

Two independent requests could run into the same code. Both see there is no existing row. So they both insert. :boom: .

While this might sound a bit theoretical, there is code that insert this missing provider state when a user connects to Nextcloud, hence there will be more than one request concurrently accessing Nextcloud.

This problem is very hard to trigger, but it's not impossible. That is why I chose the INSERT with conflict handling.

**Comment by ChristophWurst** [comment]
Also, because this might be relevant to the discussion: Two-Factor Admin is the only 2FA app that has the postgres issues. U2F and TOTP work just fine. Hence I'm really not sure if the server code is the problem here.

**Comment by henry-nicolas** [comment]
I compared the TOTP code base and the Admin 2FA and basically the StateChangeRegistryUpdater is the same. Also, the CodeStorage doesn't do anything fancy beside dispatching the event to trigger the disable of that admin provider.

I really can't wrap my head around it: updating the Dao (got the point about race conditions) makes it work (at least on Travis) while, I agree, this makes no sense since totp behave exactly as expected.

While I also doubt the root cause is on nextcloud/server, the DBAL Exception is definitvely triggered when an update is done (not an insert) for that provider.

**Comment by ChristophWurst** [comment]
> An exception occurred while executing 'UPDATE "oc_twofactor_providers"

Do we actually have more info about this? Is it always `current transaction is aborted`?

Because even if we get this error in the DAO, it's not the source of the problem. The problem is a failing operation before. The update will just error because we're using an aborted transaction there.

**Comment by henry-nicolas** [comment]
> Do we actually have more info about this? Is it always `current transaction is aborted`?

So far, yes, it is always the same error.

> Because even if we get this error in the DAO, it's not the source of the problem. The problem is a failing operation before. The update will just error because we're using an aborted transaction there.

Indeed, note that the error always pops up when using the validateCode function from CodeStorage.
Right before that, there is a codeMapper->delete call which I tried to comment, in case that would be the root cause why the transaction is aborted but no luck, the same DBAL Exception is thrown.

I could not see anything else between the codeMapper->delete call and the provider being disabled through the dispatcher.

**Comment by ChristophWurst** [comment]
> I could not see anything else between the codeMapper->delete call and the provider being disabled through the dispatcher.

That seems correct. I also have no clue where else it could fail :thinking:

**Comment by henry-nicolas** [comment]
Well, I do not understand how come the Dao works with the totp provider but not the admin one but I'm quite positive it is the root cause of this issue:
- The only SQL statement taking place before disabling the provider is the codeMapper->delete call, disabling it has no positive effect
- The update statement is preceeded in the Dao by an insert statement might not work, it looks like this is what causes the admin provider to fail

I updated the Dao persist function to avoid the race condition you described. Nice thing is that when you run an update statement the querybuilder returns the amount of update rows. This means that I basically swapped the Dao persist logic: try first an update, if no records are affected then perform an insert. The advantage is that it is clean from an SQL perspective (no primary key violation).

Not only it works with Travis CI but also on my local test infrastructure.

Please have a look at the persist function code available at:
https://github.com/henry-nicolas/server/blob/master/lib/private/Authentication/TwoFactorAuth/Db/ProviderUserAssignmentDao.php

**Comment by ChristophWurst** [comment]
> I updated the Dao persist function to avoid the race condition you described. Nice thing is that when you run an update statement the querybuilder returns the amount of updated rows. This means that I basically swapped the Dao persist logic: try first an update, if no records are affected then perform an insert. The advantage is that it is clean from an SQL perspective (no primary key violation).

Yeah, thought about this as well. But it doesn't solve the problem. It just moves it to a later stage. The problem is that you could, again, have to concurrent requests that run the same code. Both try to update and get 0 affected rows. So they insert -> one wins and the other :boom: runs into a primary key violation.

**Comment by henry-nicolas** [comment]
What about using something like this then: https://github.com/php-lock/lock

Of course it would mean adding an extra deps on nextcloud/server but it would allow the 2FA DAO to use a mutex to let us safely handle the cases you described above.

If this is not an option, I am currently out of ideas. I played quite a bit with DBAL transactions as well without success. The only success I could achieve was by avoiding the INSERT statement pkey issue.

**Comment by henry-nicolas** [comment]
Note as well that Symfony seems to have some blocking locks mechanisms as well which could be used instead of the previously mentionned library (would avoid adding an extra deps) : https://symfony.com/doc/current/components/lock.html#blocking-locks

Could be used with FlockStore for instance.

**Comment by henry-nicolas** [comment]
@ChristophWurst  let me know if you think this would be a valid approach ? Symfony locks using FlockStore. Then I would test it and we'll take it from there.

Regards,

**Comment by henry-nicolas** [comment]
Also - using the PDoStore is, IMHO, a good choice because it would work on all NC installations. The other stores are specific and we should not depend on them (Redis, MemCache, ZooKeeper) while Flock or Semaphore are local only (thus not supporting clustered setups).

Note that using PDOStore involves also adding an extra MigrationScript to get the table created to store the locks

**Comment by henry-nicolas** [comment]
@ChristophWurst - can you please give your feedback on the proposed solution (locking using PDoStore in the Dao) so that I know if it is worth investing time there?

**Comment by ChristophWurst** [comment]
Hi,

I'm generally not a fan of pessimistic locking (if I read the docs correctly and Symfony locks on a db level), as the synchronization slows down operations.

**Comment by henry-nicolas** [comment]
would you suggest any other approach ? I understand this will have a slight penalty cost but I do not see any other alternative so far

**Pull request closed without merging**